### ISSUE TITLE
SubItem is placed under wrong header after change of item

### ISSUE THREAD

**Issue by chrisonline** [davideas/FlexibleAdapter#440]
I have a list with sticky headers.

**After inital load I have two headers:**
OVERDUE
- "Test 3" with date yesterday
- "Today 8" with date yesterday

TODAY
- "Test 4" with date today
- "Test 5" with date today
- "Test 6" with date today

Now I **edit** the item "**Test 4**" and change the **date to yesterday**.
The data will be fetched again from the database and I create the list for the adapter.

The list will be added to the adapter with "**updateDataSet**" and the list **contains** this items as of my logs.:

Creating new HeaderItem: Overdue
Adding SubItem 3/Test 4 to Header Overdue
Adding SubItem 2/Test 3 to Header Overdue
Adding SubItem 7/Today 8 to Header Overdue
Adding HeaderItem to list: Overdue

Creating new HeaderItem: Today
Adding SubItem 4/Test 5 to Header Today
Adding SubItem 5/Test 6 to Header Today
Adding HeaderItem to list: Today

Reminder List contains 2 items!

Seems all **correct** and the edited item is also **updated** with the **correct date,** **BUT** the list looks now like this:
OVERDUE
- "Test 3" with date yesterday
- "Today 8" with date yesterday

TODAY
- "Test 4" with date yesterday
- "Test 5" with date today
- "Test 6" with date today

So did I miss something? Do I need a refresh of the headers or something?

For the HeaderItem I extend AbstractExpandableHeaderItem.
For the SubItems I extend AbstractSectionableItem.

If I close the app and restart it the list is again correct.

**Comment by chrisonline** [comment]
**Finally found the problem:**
In my SubItem I implemented hashCode:
The hashcode was generated via the stable ID of the item.

This was the problem because the adapter hasn't found any change because the ID haven't changed.

**Solution:**
I use as hashcode the stable id **AND date/time.**
Date/Time is changed and therefore the element is a changed for the adapter and will be recalculated.
Now it will be moved to the other header section.

Please let me know if there is a better solution.

**Comment by davideas** [comment]
@chrisonline, yes basically the RecyclerView and the Adapter work in 2 modalities for the refresh, one with _notifyDataSetChanged_ and the other with _granular notifications_.

- If you call `updateDataSet(items)` with no 2nd parameter the first mode is executed.
  - now, if you don't have stable ids every item is bound again and no animations are performed.
  - if you have initialized the adapter with stable ids, the calculation of the differences is delegated to the RecyclerView, animation should still be performed.
- if you add the 2nd parameter `true`, my calculation is executed, difference found and granular animation properly notified.

Sincerely, I still can't see advantages nor disadvantages when we use stable ids over granular notifications and viceversa. Need, to study the the RV code in that point.

Indeed, for the first case RecyclerView needs unique ids provided by items' `hashCode`. I think your solution can be acceptable, it forces the update, but what happens if you use the internal routine? Can you try? Here you have more options:
- `notifyChangeOfUnfilteredItems` already active
- and `notifyMoveOfFilteredItems` you can activate, this one calcultates if any items has changed position (to use only with max 1000 items otherwise calculation becomes too slow).

**Note:** the internal calculation is far faster than DiffUtil, especially on big numbers, that's why I deprecated the DiffUtil usage.

**PS.** You see the "Test 4" item at the bottom when you collapse because it doesn't belongs anymore to that expandable! So, the Adapter removes from the list only the subItems of that expandable and when you expand again the "father" of "Test 4" its position is restored.

**Comment by chrisonline** [comment]
@davideas What do you mean with "internal routine"?
I always call `updateDataSet(items, true)`.

My implementations of hashCode and equals was before (where the item was updated, but not moved to the new header):
```
    @Override
    public int hashCode() {
        return String.valueOf(reminderModel.get_id()).hashCode();
    }

    @Override
    public boolean equals(Object o) {
        if(o instanceof ActiveReminderItem) {
            ActiveReminderItem activeReminderItem = (ActiveReminderItem) o;
            return this.reminderModel.get_id() == activeReminderItem.reminderModel.get_id();
        }
        return false;
    }
```

**My now working soltuion is:**
```
    @Override
    public int hashCode() {
        return reminderModel.hashCode();
    }

    @Override
    public boolean equals(Object o) {
        if(o instanceof ActiveReminderItem) {
            ActiveReminderItem activeReminderItem = (ActiveReminderItem) o;
            return this.reminderModel.get_id() == activeReminderItem.reminderModel.get_id();
        }
        return false;
    }
```

But I have always called `updateDataSet(items, true)`.

**Comment by davideas** [comment]
I understood you were using `notifyDataSetChanged` with stable ids...
I tried the following test and the item is animated moved, (depends by the direction, it is moved only him or all the others in the middle).
``` java
ExpandableHeaderItem expandable1 = (ExpandableHeaderItem) DatabaseService.getInstance().getItem(1);
ExpandableHeaderItem expandable0 = (ExpandableHeaderItem) DatabaseService.getInstance().getItem(0);
expandable0.addSubItem(0, expandable1.getSubItems().remove(0));
mAdapter.setNotifyChangeOfUnfilteredItems(false);
mAdapter.setNotifyMoveOfFilteredItems(true);
mAdapter.updateDataSet(DatabaseService.getInstance().getDatabaseList(), true);
```
- That is a an automatic calculation done in background without using stableIds and it works.
- I also tried using the stable ids and `updateDataSet(newItems, false)`: it also works, item is animated, and no time value in the hashCode! I think you first need to change the original list as I did it by removing the subItems from a header and add it to another (execution is very fast).
- Another solution is to give granular instructions to the Adapter, after you change the database items, saying: this item at position `x` is now at position `y` using `moveItem()`. Of course you need to calculate the positions `x, y`.

**Comment by chrisonline** [comment]
OK thanks for the info. To remove and add the subitem I have to completly recode the part because now I don't know which item is changed.
I use a Cursorloader and after finish of the loader I readd all items from the cursor to the list and call `updateDataSet(newItems, true)`!

But I know what you mean and I will check if I change it again.

One more question about the methods `equals()` and `hashcode()`:
As my understanding the method `equals()` is not used internally. I can use it and check if the object is the same (not if there is a title change or whatever, only if this is the same ID).
The method `hashcode()` is used to check if an item is changed. So for example if the ID is the same but the title has changed.
So the hashcode has to be build for the complete item variables/values.

Is this correct?

**Comment by davideas** [comment]
@chrisonline, `equals` is used by Collecitons based on List, `hashCode` by Collections based on Hash, necessary to identify the item during the scan to pick it up / remove / update etc... But this is Java concept.
In the Adapter both collections are used.

I've added a method `IFlexible.shouldNotifyChange(new Item)` (similar to DiffUtil concept) that checks if the Item must be really updated or skipped, inside there you can evaluate all the others fields. By default it returns `true`.

The cursor should return only the Expandables as per requirement explained in the Wiki _headers and sections_. So you have to move the subItem from the subList to the other subList before providing the newList of only expandables to the adapter.
